### PR TITLE
Systemd service should use /bin/karaf instead of start.sh

### LINKED ISSUE #70: Better handling of openHAB startup problems in systemd
Hey all, Hey @BClark09,

we are seeing problems similar to [this](https://community.openhab.org/t/openhab-cannot-start-after-upgrade-ports-not-opened/31527) quite often.

To the extend of your knowledge, can you think of any easy or practical way to have `systemctl status` actually report startup failures? It might be worth it to look into this topic. Best! Thomas

Also see: https://github.com/openhab/openhabian/issues/50#issuecomment-280868594

### PR DESCRIPTION
**Pull request by BClark09**
Closes #70 

Signed-off-by: Ben Clark <ben@benjyc.uk>

### CODE REVIEW

**Comment by BClark09** [comment]
@ThomDietrich if it helps for testing, the above changes are available in the linuxpkg-testing repo:

```
echo 'deb https://openhab.jfrog.io/openhab/linuxpkg-testing unstable main' | sudo tee /etc/apt/sources.list.d/openhab2.list
```

**Comment by BClark09** [comment]
```
OPENHAB_USERDATA=WHEREAMI
```

Before PR:
image

After PR:
image

Seems to have gone well!

**Comment by ThomDietrich** [comment]
Excellent! I'm sadly not able to do any tests today. I'm curious how the error would look like if there is no java runtime available

**Comment by ThomDietrich** [comment]
One bonus question: The current result is "exit-code". Do you want to spend extra time and effort to define a few known fail reasons? This is something I could look into this weekend but feel free to tinker around as well ;)

**Comment by BClark09** [comment]
Yes, but there's a weird issue with this: 

When a program managed by systemd sources a script, the stout buffer from that script seems delayed. In this case, when karaf calls our `setenv` script the `echo 'Java version higher than 1.8...'` does not get the chance to print when `exit 1;` follows. 

To see this error message in systemd, we would have to edit 'setenv' to output with:

```
echo "Error message" && /bin/true
```

**Comment by ThomDietrich** [comment]
@BClark09 your last comment confused me. Why will echo be "slower" than any following command? If it is like you suggested and stdout is cut off to early at the exit, did you find a discussion on the issue? This one looks alright: http://systemd-devel.freedesktop.narkive.com/suwZf7s7/unbuffered-stderr-for-my-systemd-service

Your proposed solution does work because the otherwise useless `&& /bin/true` will delay following commands? Is there a cleaner way to do that? sd_notify would be an option for systemd. Another option might be sleep but that's hack'y...

**Comment by BClark09** [comment]
There are a few scattered around [[1]](https://serverfault.com/questions/849249/turn-off-buffering-in-systemd-unit-file), I think [this is the systemd bug](https://github.com/systemd/systemd/issues/2913) to watch for this issue though.

**Comment by BClark09** [comment]
I think this is okay to merge now. I'll wait a bit for feedback in the above thread to see if the ulimit restriction has been adjusted.

File: resources/systemd/openhab2.service [hunk 1]
```diff
@@ -1,5 +1,5 @@
 [Unit]
-Description=openHAB 2 - empowering the smart home
+Description=openHAB 2 - Empowering the smart home
```
Comment (ThomDietrich) [inline resources/systemd/openhab2.service]: Why?

File: resources/etc/default/openhab2 [hunk 2]
```diff
@@ -29,3 +29,8 @@ EXTRA_JAVA_OPTS=""
 #
 #OPENHAB_USER=openhab
 #OPENHAB_GROUP=openhab
+#
+# The startmode for the openHAB runtime. Only available for systemctl/systemd systems.
+# Defaults to daemon when unset here.
```
Comment (ThomDietrich) [inline resources/etc/default/openhab2]: Should we list the options here?

File: resources/systemd/openhab2.service (hunk 1 above)
Comment (BClark09) [inline resources/systemd/openhab2.service]: It seemed un-natural at it was. My intention was to capitalise the S and H in Smart Home (as it is in the docs), but I must have misplaced that.

Comment (ThomDietrich) [inline resources/systemd/openhab2.service]: I also wanted to do that a few months ago but got scolded by Kai for it.

http://docs.openhab.org/images/logo.png

Comment (ThomDietrich) [inline resources/systemd/openhab2.service]: https://github.com/openhab/openhab-distro/pull/271#r77364735

File: resources/etc/default/openhab2 (hunk 2 above)
Comment (BClark09) [inline resources/etc/default/openhab2]: I wasn't so sure, options for karaf are:

```
clean   - Deletes $KARAF_DATA contents
debug   - Turns on java debugging options
status  - Checks if running
stop    - Stops instance
console - ???
server  - Starts remote shell and runs the karaf server
daemon  - Same as server, extra exec and return details.
client  - Starts the client to an existing instance
```

Out of them, we could really only recommend using `daemon`, `debug` or `debug daemon` with systemd.

File: resources/systemd/openhab2.service (hunk 1 above)
Comment (BClark09) [inline resources/systemd/openhab2.service]: I see... I'll change it back then.

File: resources/etc/default/openhab2 (hunk 2 above)
Comment (ThomDietrich) [inline resources/etc/default/openhab2]: Agreed.

**Comment by BClark09** [comment]
@kaikreuzer, @theoweiss, just an FYI: I noticed there was some discussion at https://github.com/openhab/openhab-distro/pull/138 about how to start using systemd. I have changed this to use `runtime/bin/karaf daemon` so that systemd recognises java as the main process to follow. It is also the recommended way using the [karaf service template](https://github.com/apache/karaf/blob/master/assemblies/features/base/src/main/resources/resources/bin/contrib/karaf-service-template.systemd) by the looks of things. 

Is there any reason, or can you remember any reason for this being a bad idea? Tests on Debian and Fedora have worked well.

**Comment by kaikreuzer** [comment]
I am not aware of reason against it, but @theoweiss is indeed probably the much better person to ask about it :-)

**Comment by ThomDietrich** [comment]
@BClark09 Instead of the `&& /bin/true` trick we still could use `systemd-notify`/`sd_notify`. Did you look into them?

**Comment by BClark09** [comment]
I could find a way of incorporating it reliably, it seems to suffer from the same race condition.

Also, should we really add platform specific Linux commands to openHAB's start scripts?

**Comment by ThomDietrich** [comment]
I thought about it and couldn't see a reason why not. Of course one would need to properly check for systemd beforehand.

>  it seems to suffer from the same race condition

Well nice...

**Comment by BClark09** [comment]
Actually, I may have been wrong in that, was looking in the wrong place. The systemd service needed a bit of tweaking to be of "notify" type.

image

Unfortunately, this still isn't a solution, openHAB would need an implementation of sd_notify to pass down that the service has started successfully. Otherwise, the service start command would keep waiting for one and timeout.

**Comment by ThomDietrich** [comment]
You mean this would need to happen inside openHAB or Karaf, not inside the bash script!?

**Comment by BClark09** [comment]
If you're using sd_notify, then you'll need to pass down a message that the service has started successfully, which should only happen inside the java process started by Karaf if you want to certain that everything is working correctly.

**Comment by BClark09** [comment]
To summarise:

- This PR makes sure that **some** information is available.
- Additionally, the service now has a large file limit as suggested by the Karaf template.
- Since I've re-enabled automatic restarting (with some delay now) these messages only appear when the user uses `journalctl -u openhab2 -b` (report all openhab2 service messages since last boot). This is what we should suggest to users for debugging.
- The service still requires some editing of `setenv` to obtain all output.

I'd suggest we merge now and I can make a PR in the distro if using ` && /bin/true` is an acceptable workaround.

**Comment by theoweiss** [comment]
Sorry guys for not participating in this discussion but remodelling our new house and working for my paid  job is still eating my whole time. May be I'm back in late autumn. @BClark09 is doing a great job here!!

**Comment by BClark09** [comment]
No problem @theoweiss! Hope it's all going well!

**Pull request merged**